Bugfix/sim 1562/speed up query columns

danielsf opened the pull request:
This pull request deals with one of the slowdowns we experienced due to numpy 1.10 (see the associated JIRA ticket for profiling data).

I also took the opportunity to modernize some of the unit test code.

danielsf committed:
do not convert query results to a recarray until the last possible minute.
This achieves a factor of ~2 speed-up under numpy 1.10 (when recarray
access via field names slowed down by at least two orders of magnitude)

danielsf committed:
replace os.getenv with lss.utils.getPackageDir

danielsf committed:
remove comment about expanding unit tests with CATSIM-90

that issue is now SIM-194 and is marked as Done

danielsf committed:
added unit test to make sure that default values are properly applied
in CatalogDBObject.query_columns

danielsf committed:
use with statement to open data file

jbkalmbach commented on python/lsst/sims/catalogs/generation/db/dbConnection.py:
```diff
@@ -542,24 +542,31 @@ def _postprocess_results(self, results):
-                            raise TypeError("Couldn't convert column %s"%(k))
+
+            for colName in cols:
+                if colName not in results[0].keys() and colName not in self.dbDefaultValues:
```
Isn't colName in results[0].keys() by definition of cols in line 547?

danielsf replied:
Yes.  I'll be honest: I am not actually sure how that TypeError could ever have been raised, even before this pull request.

I assumed it was there for a reason.

Maybe it was there in case there was a type mismatch between results and retresults.

Regardless, I will drop this check (lines 555 through 557).

jbkalmbach commented on tests/testCatalogDBObject.py:
```diff
@@ -35,20 +36,40 @@ def createNonsenseDB():
 
     conn.commit()
     conn.close()
     inFile.close()
```
If you use 'with', can't you remove this step?

danielsf replied:
ah. yes.

jbkalmbach commented:
Looks good.

danielsf committed:
removed check that could never fail (colName was defined to be in
results.keys() a few lines earlier with

cols = [str(k) for k in results[0].keys()]

danielsf committed:
removed unnecessary inFile.close() (we are now using a with statement)
